### PR TITLE
[osdp_fixes 2/5] Various PD cleanups, enhancements and fixes

### PR DESCRIPTION
**Pull request by sidcha**
This PR brings in various updates such as bug fixes, enhancements and cleanups to subsys/mgmt/osdp in peripheral device mode. These changes have been tested with a st-nucleo-g474re board with secure channel enabled.

This PR is a split version of #46449 to improve review-ability (changes remain the same).

### CODE REVIEW

File: subsys/mgmt/osdp/src/osdp_pd.c [hunk 1]
```diff
@@ -189,6 +189,69 @@ static bool do_command_callback(struct osdp_pd *pd, struct osdp_cmd *cmd)
...
+			pd->reply_id = REPLY_NAK;
+			pd->ephemeral_data[0] = OSDP_PD_NAK_SC_UNSUP;
+			return 0;
+		}
+		return 1;
+	}
+
+	pd->reply_id = REPLY_NAK;
+	pd->ephemeral_data[0] = OSDP_PD_NAK_CMD_UNKNOWN;
+	LOG_INF("PD is not capable of handling CMD(%02x); "
```
Comment (gmarull) [inline subsys/mgmt/osdp/src/osdp_pd.c:253]: LOG_INF feels wrong here, maybe error/warn/dbg?

**Review by gmarull — DISMISSED** [review]
lgtm, one minor nitpick

File: subsys/mgmt/osdp/src/osdp_pd.c [hunk 2]
```diff
@@ -226,13 +292,21 @@ static int pd_decode_command(struct osdp_pd *pd, uint8_t *buf, int len)
 		if (len != CMD_ISTAT_DATA_LEN) {
 			break;
 		}
+		if (!pd_cmd_cap_ok(pd, NULL)) {
```
Comment (adakus) [inline subsys/mgmt/osdp/src/osdp_pd.c:295]: Are there any requirements defined for POLL command ?

File: subsys/mgmt/osdp/src/osdp_pd.c [hunk 3]
```diff
@@ -380,22 +466,26 @@ static int pd_decode_command(struct osdp_pd *pd, uint8_t *buf, int len)
 		if (len != CMD_KEYSET_DATA_LEN) {
 			break;
 		}
+		/* only key_type == 1 (SCBK) and key_len == 16 is supported */
```
Comment (adakus) [inline subsys/mgmt/osdp/src/osdp_pd.c:469]: Does spec defines any limitation for key length  ?

File: subsys/mgmt/osdp/src/osdp_phy.c [hunk 4]
```diff
@@ -131,16 +161,33 @@ int osdp_phy_packet_finalize(struct osdp_pd *pd, uint8_t *buf,
 	uint16_t crc16;
 	struct osdp_packet_header *pkt;
 
-	/* Do a sanity check only as we expect expect header to be prefilled */
-	if (buf[0] != OSDP_PKT_MARK || buf[1] != OSDP_PKT_SOM) {
+	/* Do a sanity check only; we expect expect header to be pre-filled */
```
Comment (adakus) [inline subsys/mgmt/osdp/src/osdp_phy.c:164]: expect expect

File: subsys/mgmt/osdp/src/osdp_pd.c [hunk 5]
```diff
@@ -635,6 +639,46 @@ static int pd_build_reply(struct osdp_pd *pd, uint8_t *buf, int max_len)
...
+		len_bytes = (event->cardread.length + 7) / 8;
+		assert_buf_len(REPLY_RAW_LEN + len_bytes, max_len);
+		buf[len++] = pd->reply_id;
+		buf[len++] = (uint8_t)event->cardread.reader_no;
+		buf[len++] = (uint8_t)event->cardread.format;
+		buf[len++] = BYTE_0(event->cardread.length);
+		buf[len++] = BYTE_1(event->cardread.length);
+		for (i = 0; i < len_bytes; i++) {
+			buf[len++] = event->cardread.data[i];
+		}
```
Comment (r2r0) [inline subsys/mgmt/osdp/src/osdp_pd.c:671]: Consider usnig `memcpy` for data copying.

File: subsys/mgmt/osdp/src/osdp_pd.c [hunk 6]
```diff
@@ -269,6 +340,9 @@ static int pd_decode_command(struct osdp_pd *pd, uint8_t *buf, int len)
 		cmd.output.timer_count = buf[pos++];
 		cmd.output.timer_count |= buf[pos++] << 8;
 		ret = OSDP_PD_ERR_REPLY;
+		if (!pd_cmd_cap_ok(pd, &cmd)) {
+			break;
+		}
```
Comment (r2r0) [inline subsys/mgmt/osdp/src/osdp_pd.c:348]: Move capablity checking before setting any reply information.
It applies here and for CMD_LED, CMD_BUZ and CMD_TEXT.

File: subsys/mgmt/osdp/src/osdp_pd.c (hunk 1 above)
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:253]: Yes, changing this to an error level log.

File: subsys/mgmt/osdp/src/osdp_pd.c (hunk 2 above)
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:295]: No, all PD have to support the POLL command.

File: subsys/mgmt/osdp/src/osdp_pd.c (hunk 3 above)
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:469]: The spec defines the key size and block size to be 128 bits that give us the 16 bytes.

File: subsys/mgmt/osdp/src/osdp_phy.c (hunk 4 above)
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_phy.c:164]: Ack, thanks.

File: subsys/mgmt/osdp/src/osdp_pd.c (hunk 5 above)
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:671]: Ack, thanks.

File: subsys/mgmt/osdp/src/osdp_pd.c (hunk 6 above)
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:348]: Not sure I follow. Which reply information are you taking about?

Comment (r2r0) [reply subsys/mgmt/osdp/src/osdp_pd.c:348]: I.e. in this case `if (!pd_cmd_cap_ok(pd, &cmd)) {...}` should be moved just after `if (len != CMD_OUT_DATA_LEN || !pd->command_callback) {...}`

Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:348]: Ah, that. No we cannot do that; the cap check needs the command that we decoded in som cases (wherever cmd arg is non-null).

File: subsys/mgmt/osdp/src/osdp_pd.c (hunk 3 above)
Comment (r2r0) [reply subsys/mgmt/osdp/src/osdp_pd.c:469]: In the 2.1.7 there is stated: "The Length indicates the number of bytes containing key data in the array Data[]. It is
computed as the integer value of the quantity of Key Length in bits plus 7 divided by 8. For
example, the Length shall be 16, and the Data[] array shall contain the 128-bit Secure
Channel base key (SCBK). For the 256-bit key the Length shall be 32."
Also in the 2.2 spec does not state exact 16 byte key length.

File: subsys/mgmt/osdp/src/osdp_pd.c (hunk 2 above)
Comment (adakus) [reply subsys/mgmt/osdp/src/osdp_pd.c:295]: So you can remove this capabilities check

File: subsys/mgmt/osdp/src/osdp_pd.c (hunk 3 above)
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:469]: > In the 2.1.7 there is stated: "The Length indicates the number of bytes containing key data in the array Data[]. It is
computed as the integer value of the quantity of Key Length in bits plus 7 divided by 8. For
example, the Length shall be 16, and the Data[] array shall contain the 128-bit Secure
Channel base key (SCBK). 

Yes, I am aware of this. A lot of parts of mgmt/osdp is designed with the assumption `key_size = block_size = 16` (for that matter even the spec, see last point). If that were to change in a future version (right now there isn't any), then we need to take a look all over for such assumptions and fix them. For now, this check exists to assert it.

> For the 256-bit key the Length shall be 32.

I don't have 2.1.7, I think this has been removed in the latest version 2.2 of the spec (I cannot find it in 2.15 too).

> Also in the 2.2 spec does not state exact 16 byte key length

Not directly, no. But the spec make that obvious in various places. For instance,
- the SCBK default key is defined for 128 bits only.
- the capability mask defines only a 128 bit key.
- session key, MAC, derivation process talks about 128 bit keys only.

File: subsys/mgmt/osdp/src/osdp_pd.c (hunk 2 above)
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:295]: This check is not in `case CMD_POLL`. It is for `case CMD_ISTAT`.

**Comment by sidcha** [comment]
@gmarull, Since your last review, 2 new (trivial) commits are added to this PR:
- 084841aa78 mgmt/osdp: Use memcpy instead of raw loops in many places
- 9ac2e5668d mgmt/osdp: pd: Fix device capabilities report

@adakus, @r2r0 I believe I have addressed all your comments

Please take another look.

File: subsys/mgmt/osdp/src/osdp_phy.c [hunk 7]
```diff
@@ -385,6 +385,12 @@ int osdp_phy_decode_packet(struct osdp_pd *pd, uint8_t *buf, int len)
 			pd->ephemeral_data[0] = OSDP_PD_NAK_SC_COND;
 			return OSDP_ERR_PKT_FMT;
 		}
+		if (!sc_is_active(pd) && pkt->data[1] > SCS_14) {
+			LOG_ERR("Received invalid secure message!");
```
Comment (r2r0) [inline subsys/mgmt/osdp/src/osdp_phy.c:387]: For CMD_SCRYPT there is LOG_WRN and here is LOG_ERR.
Both are cases where command is NACKed.

Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_phy.c:387]: Should have been an error level log too. Fixed.

**Review by r2r0 — DISMISSED** [review]

**Review by adakus — DISMISSED** [review]

**Review by r2r0 — APPROVED** [review]

**Review by adakus — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]

**Comment by stephanosio** [comment]
@r2r0 @adakus It seems like you have not accepted the zephyrproject-rtos GitHub organisation invitation and your reviews are not counted towards the approvals required for merging this PR because of that (note that your approvals currently show up as grey). Please accept the invitation so that your reviews count and this PR can be merged.

**Pull request merged**